ValveResourceFormat/Resource/Blocks/SNAP.cs

```diff
@@ -12,16 +16,138 @@ public class SNAP : Block
     {
         public override BlockType Type => BlockType.SNAP;
 
-        public override void Read(BinaryReader reader, Resource resource)
-        {
-            reader.BaseStream.Position = Offset;
+        public uint NumParticles { get; private set; }
 
-            throw new NotImplementedException();
-        }
+        public IReadOnlyDictionary<string, IEnumerable> AttributeData { get; private set; }
 
         public override void WriteText(IndentedTextWriter writer)
         {
             writer.WriteLine("{0:X8}", Offset);
+
+            writer.WriteLine($"{NumParticles} particles with {AttributeData.Count} attributes:");
+            writer.WriteLine();
+
+            foreach (var (attribute, data) in AttributeData)
+            {
+                writer.WriteLine($"- Attribute {attribute} -");
+                foreach (var d in data)
+                {
+                    writer.WriteLine(d);
+                }
+                writer.WriteLine();
+            }
+        }
+
+        public override void Read(BinaryReader resourceReader, Resource resource)
+        {
+            resourceReader.BaseStream.Position = Offset;
+
+            // Decompress SNAP block compression
+            using var decompressedStream = BlockCompress.Decompress(resourceReader, Size);
+            using var reader = new BinaryReader(decompressedStream);
+
+            // Get DATA block to know how to read SNAP data
+            var data = resource.DataBlock.AsKeyValueCollection();
+
+            var numParticles = data.GetIntegerProperty("num_particles");
+            var attributes = data.GetArray("attributes");
+            var stringList = data.GetArray<string>("string_list");
+
+            var attributeData = new Dictionary<string, IEnumerable>();
+
+            foreach (var attribute in attributes)
+            {
+                var attributeName = attribute.GetProperty<string>("name");
+                var attributeType = attribute.GetProperty<string>("type");
+
+                var attributeArray = attributeType switch
```

Comment:
From IDA, the possible types should be: int, float, float3, skinning, vector, string, bone

So missing int and bone right now.